packages/doenetml-worker-rust/lib-doenetml-core/src/core/parse_action.rs

```diff
@@ -0,0 +1,59 @@
+use anyhow::Result;
+
+use crate::components::doenet::text_input::TextInputAction;
+use crate::components::ComponentEnum;
+use crate::{AllActions, ComponentIdx, DoenetMLCore};
+
+#[derive(Debug, serde::Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct ActionWithOnlyComponentIndex {
+    component_idx: usize,
+}
+
+impl DoenetMLCore {
+    /// Convert a json string into an action. The action must contain a `componentIdx` field.
+    /// The type of `self.components[componentIdx]` is checked and the action deserialization
+    /// will fail if it is not compatible with the component.
+    pub fn deserialize_action(
+        &self,
+        action_json: &str,
+    ) -> Result<(ComponentIdx, AllActions), String> {
+        let action_with_only_component_index: ActionWithOnlyComponentIndex =
+            serde_json::from_str(action_json).map_err(|e| {
+                format!(
+                    "Cannot find `componentIdx on action `{}`. Got error: {}",
+                    action_json, e
+                )
+            })?;
+        let idx = action_with_only_component_index.component_idx;
+        let component = self
+            .components
+            .get(idx)
+            .ok_or_else(|| format!("Cannot find component with index/id {}", idx))?;
+
+        let make_error_string = |e: serde_json::Error| {
+            format!(
+                "Cannot deserialize action `{}` for component `{}`. Got error: {}",
+                action_json,
+                component.borrow(),
+                e
+            )
+        };
+
+        let action = match *component.borrow() {
+            ComponentEnum::TextInput(_) => {
```

Comment:
Can we write this so that we don't have to hard code any particular component type in this file? I'd like to make it easy as easy as possible to add new components, where it is just a matter of creating a new file `components/doenet` and then adding to enums in `cmponents/common.rs`.

I would suggest adding this logic to a trait of `RenderedComponentNode`, so that the implementation could be coded in each component file. The `enum_dispatch` macro will then automatically implement the trait on `ComponentEnum` so that it could be called here as something like `component.deserialize_action(action_json)`. (Not sure if there is common preprocessing of `action_json` before sending to the component's implementation.)

Reply:
This has been eliminated by putting a `component` field in an `Action`